File: youtube/lat9-cancer/try1.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns 
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns 
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report
from sklearn.metrics import confusion_matrix
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.neural_network import MLPClassifier
from sklearn.neural_network import MLPRegressor
from sklearn.preprocessing import MinMaxScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.neural_network import MLPClassifier
from sklearn.neural_network import MLPRegressor

df = pd.read_csv('ML/youtube/lat9-cancer/data.csv')
print(df.head())
print(df.info())
print(df.describe())
print(df.isnull().sum())

df = df.drop('id', axis = 1)
del df['Unnamed: 32']

for col in df:
    print(col)
    print(df[col].nunique())
    print()

print(df['diagnosis'].unique())
df['diagnosis'] = df['diagnosis'].map({'M': 0, 'B': 1})


fig, ax = plt.subplots(6, 5, figsize=(20, 20))
# Menggunakan range dari 0 sampai 5 untuk iterasi setiap kolom dalam DataFrame
sns.scatterplot(x=df.index, y=df['diagnosis'], ax=ax[0, 0], color='blue')
sns.scatterplot(x=df.index, y=df['radius_mean'], ax=ax[0, 1], color='green')
sns.scatterplot(x=df.index, y=df['texture_mean'], ax=ax[0, 2], color='orange')
sns.scatterplot(x=df.index, y=df['perimeter_mean'], ax=ax[0, 3], color='red')
sns.scatterplot(x=df.index, y=df['area_mean'], ax=ax[0, 4], color='blue')
sns.scatterplot(x=df.index, y=df['smoothness_mean'], ax=ax[1, 0], color='green')
sns.scatterplot(x=df.index, y=df['compactness_mean'], ax=ax[1, 1], color='orange')
sns.scatterplot(x=df.index, y=df['concavity_mean'], ax=ax[1, 2], color='red')
sns.scatterplot(x=df.index, y=df['concave points_mean'], ax=ax[1, 3], color='blue')
sns.scatterplot(x=df.index, y=df['symmetry_mean'], ax=ax[1, 4], color='green')
sns.scatterplot(x=df.index, y=df['fractal_dimension_mean'], ax=ax[2, 0], color='orange')
sns.scatterplot(x=df.index, y=df['radius_se'], ax=ax[2, 1], color='red')
sns.scatterplot(x=df.index, y=df['texture_se'], ax=ax[2, 2], color='blue')
sns.scatterplot(x=df.index, y=df['perimeter_se'], ax=ax[2, 3], color='green')
sns.scatterplot(x=df.index, y=df['area_se'], ax=ax[2, 4], color='orange')
sns.scatterplot(x=df.index, y=df['smoothness_se'], ax=ax[3, 0], color='red')
sns.scatterplot(x=df.index, y=df['compactness_se'], ax=ax[3, 1], color='blue')
sns.scatterplot(x=df.index, y=df['concavity_se'], ax=ax[3, 2], color='green')
sns.scatterplot(x=df.index, y=df['concave points_se'], ax=ax[3, 3], color='orange')
sns.scatterplot(x=df.index, y=df['symmetry_se'], ax=ax[3, 4], color='red')
sns.scatterplot(x=df.index, y=df['fractal_dimension_se'], ax=ax[4, 0], color='blue')
sns.scatterplot(x=df.index, y=df['radius_worst'], ax=ax[4, 1], color='green')
sns.scatterplot(x=df.index, y=df['texture_worst'], ax=ax[4, 2], color='orange')
sns.scatterplot(x=df.index, y=df['perimeter_worst'], ax=ax[4, 3], color='red')
sns.scatterplot(x=df.index, y=df['area_worst'], ax=ax[4, 4], color='blue')
sns.scatterplot(x=df.index, y=df['smoothness_worst'], ax=ax[5, 0], color='green')
sns.scatterplot(x=df.index, y=df['compactness_worst'], ax=ax[5, 1], color='orange')
sns.scatterplot(x=df.index, y=df['concavity_worst'], ax=ax[5, 2], color='red')
sns.scatterplot(x=df.index, y=df['concave points_worst'], ax=ax[5, 3], color='blue')
sns.scatterplot(x=df.index, y=df['symmetry_worst'], ax=ax[5, 4], color='green')
plt.tight_layout()
plt.show()

def remove_outlier(df, col):
    # q1 = np.percentile(df[col], 25)
    # q3 = np.percentile(df[col], 75)
    # iqr = q3 - q1
    # df = df[(df[col] >= q1 - 1.5*iqr) & (df[col] <= q3 + 1.5*iqr )]
    
    q1 = df[col].quantile(0.25)
    q3 = df[col].quantile(0.75)
    iqr = q3 - q1
    df = df[(df[col] >= q1 - 1.5*iqr) & (df[col] <= q3 + 1.5*iqr )]
    return df


for col in df:
    df = remove_outlier(df, col)

fig, ax = plt.subplots(6, 5, figsize=(20, 20))
# Menggunakan range dari 0 sampai 5 untuk iterasi setiap kolom dalam DataFrame
sns.scatterplot(x=df.index, y=df['diagnosis'], ax=ax[0, 0], color='blue')
sns.scatterplot(x=df.index, y=df['radius_mean'], ax=ax[0, 1], color='green')
sns.scatterplot(x=df.index, y=df['texture_mean'], ax=ax[0, 2], color='orange')
sns.scatterplot(x=df.index, y=df['perimeter_mean'], ax=ax[0, 3], color='red')
sns.scatterplot(x=df.index, y=df['area_mean'], ax=ax[0, 4], color='blue')
sns.scatterplot(x=df.index, y=df['smoothness_mean'], ax=ax[1, 0], color='green')
sns.scatterplot(x=df.index, y=df['compactness_mean'], ax=ax[1, 1], color='orange')
sns.scatterplot(x=df.index, y=df['concavity_mean'], ax=ax[1, 2], color='red')
sns.scatterplot(x=df.index, y=df['concave points_mean'], ax=ax[1, 3], color='blue')
sns.scatterplot(x=df.index, y=df['symmetry_mean'], ax=ax[1, 4], color='green')
sns.scatterplot(x=df.index, y=df['fractal_dimension_mean'], ax=ax[2, 0], color='orange')
sns.scatterplot(x=df.index, y=df['radius_se'], ax=ax[2, 1], color='red')
sns.scatterplot(x=df.index, y=df['texture_se'], ax=ax[2, 2], color='blue')
sns.scatterplot(x=df.index, y=df['perimeter_se'], ax=ax[2, 3], color='green')
sns.scatterplot(x=df.index, y=df['area_se'], ax=ax[2, 4], color='orange')
sns.scatterplot(x=df.index, y=df['smoothness_se'], ax=ax[3, 0], color='red')
sns.scatterplot(x=df.index, y=df['compactness_se'], ax=ax[3, 1], color='blue')
sns.scatterplot(x=df.index, y=df['concavity_se'], ax=ax[3, 2], color='green')
sns.scatterplot(x=df.index, y=df['concave points_se'], ax=ax[3, 3], color='orange')
sns.scatterplot(x=df.index, y=df['symmetry_se'], ax=ax[3, 4], color='red')
sns.scatterplot(x=df.index, y=df['fractal_dimension_se'], ax=ax[4, 0], color='blue')
sns.scatterplot(x=df.index, y=df['radius_worst'], ax=ax[4, 1], color='green')
sns.scatterplot(x=df.index, y=df['texture_worst'], ax=ax[4, 2], color='orange')
sns.scatterplot(x=df.index, y=df['perimeter_worst'], ax=ax[4, 3], color='red')
sns.scatterplot(x=df.index, y=df['area_worst'], ax=ax[4, 4], color='blue')
sns.scatterplot(x=df.index, y=df['smoothness_worst'], ax=ax[5, 0], color='green')
sns.scatterplot(x=df.index, y=df['compactness_worst'], ax=ax[5, 1], color='orange')
sns.scatterplot(x=df.index, y=df['concavity_worst'], ax=ax[5, 2], color='red')
sns.scatterplot(x=df.index, y=df['concave points_worst'], ax=ax[5, 3], color='blue')
sns.scatterplot(x=df.index, y=df['symmetry_worst'], ax=ax[5, 4], color='green')
plt.tight_layout()
plt.show()


print(df.head())
numvars = df.drop('diagnosis', axis = 1).columns
scaler = MinMaxScaler()
df[numvars] = scaler.fit_transform(df[numvars])
print(df.head())

x = df
y = df.pop('diagnosis')


X_train, X_test, y_train, y_test = train_test_split(x,y, test_size=0.3, random_state=42, stratify=y)
print(y_train.value_counts(normalize = True))
print(y_test.value_counts(normalize = True))


k_folds = KFold(n_splits=5, shuffle=True, random_state=42)
modelLR  = LogisticRegression()
scoring = 'accuracy'
score = cross_val_score(modelLR, X_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))



modelRF  = RandomForestClassifier()
scoring = 'accuracy'
score = cross_val_score(modelRF, X_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))



modelMPL  = MLPClassifier()
scoring = 'accuracy'
score = cross_val_score(modelMPL, X_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))


modelNB  = GaussianNB()
scoring = 'accuracy'
score = cross_val_score(modelNB, X_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))

modelLR.fit(X_train, y_train)
y_pred_train = modelLR.predict(X_train)

train = pd.DataFrame({
    'y_train': y_train,
    'y_pred_train': y_pred_train
})
print(train)
print(train[train['y_train'] != train['y_pred_train']])

akurasi = accuracy_score(y_train, y_pred_train)
print(akurasi)
cm = confusion_matrix(y_train, y_pred_train)
print(cm)
sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
plt.xlabel('y_pred_train')
plt.ylabel('y_train')
plt.title('confusion matrix')
plt.show()


y_pred_test = modelLR.predict(X_test)

train = pd.DataFrame({
    'y_test': y_test,
    'y_pred_test': y_pred_test
})
print(train)
print(train[train['y_test'] != train['y_pred_test']])

akurasi = accuracy_score(y_test, y_pred_test)
print(akurasi)
cm = confusion_matrix(y_test, y_pred_test)
print(cm)
sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
plt.xlabel('y_pred_test')
plt.ylabel('y_test')
plt.title('confusion matrix')
plt.show()

File: youtube/lat9-cancer/try1.r
library(dplyr)
library(ggplot2)
library(stringr)
library(gridExtra)
library(RMySQL)
library(dplyr)
library(caTools)
library(tidyr)
library(plyr)
library(glmnet)
library(lubridate)
library(tidyverse)
library(randomForest)
library(caret)
library(patchwork)
library(nnet)
library(klaR) # untuk Naive Bayes
library(rpart) # untuk Decision Trees


df <- read.csv('ML/youtube/lat9-cancer/data.csv')
print(head(df))
print(str(df))
print(summary(df))
print(colSums(is.na(df)))

df <- dplyr::select(df, -id, -X)

for (col in colnames(df)){
    print(col)
    print(head(unique(df[col])))
    cat('\n\n')
}



print(sapply(df, function(x) length(unique(na.omit(x)))))
# print(sapply(df, function(x) length(unique(x[!is.na(x)]))))


print(unique(df$diagnosis))

# df$diagnosis <- ifelse(df$diagnosis == 'M', 0, 1)
df$diagnosis <- case_when(
    df$diagnosis == 'M' ~ 0,
    .default = 1
)
print(distinct(df, diagnosis))


# Membuat plot untuk setiap kolom
p1 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = diagnosis), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p2 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = radius_mean), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p3 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = texture_mean), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p4 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = perimeter_mean), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p5 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = area_mean), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p6 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = smoothness_mean), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p7 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = compactness_mean), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p8 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concavity_mean), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p9 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concave.points_mean), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p10 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = symmetry_mean), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p11 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fractal_dimension_mean), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p12 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = radius_se), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p13 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = texture_se), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p14 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = perimeter_se), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p15 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = area_se), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p16 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = smoothness_se), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p17 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = compactness_se), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p18 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concavity_se), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p19 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concave.points_se), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p20 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = symmetry_se), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p21 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fractal_dimension_se), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p22 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = radius_worst), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p23 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = texture_worst), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p24 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = perimeter_worst), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p25 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = area_worst), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p26 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = smoothness_worst), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p27 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = compactness_worst), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p28 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concavity_worst), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p29 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concave.points_worst), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p30 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = symmetry_worst), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p31 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fractal_dimension_worst), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()

# Menggabungkan plot menggunakan patchwork
plot <- p1 + p2 + p3 + p4 + p5 + p6 + p7 + p8 + p9 + p10 + 
        p11 + p12 + p13 + p14 + p15 + p16 + p17 + p18 + p19 + p20 + 
        p21 + p22 + p23 + p24 + p25 + p26 + p27 + p28 + p29 + p30 + p31
print(plot)


remove_outlier <- function(df, col){
    # q1 <- quantile(df[[col]], 0.25)
    # q3 <- quantile(df[[col]], 0.75)
    # iqr <- q3 - q1
    # # Mengaplikasikan aturan IQR untuk menyaring data
    # df <- subset(df, df[[col]] >= q1 - 1.5 * iqr & df[[col]] <= q3 + 1.5 * iqr)


    q1 <- quantile(df[[col]], 0.25)
    q3 <- quantile(df[[col]], 0.75)
    iqr <- q3 - q1
    # Mengaplikasikan aturan IQR untuk menyaring data
    df <- filter(df, df[[col]] >= q1 - 1.5 * iqr & df[[col]] <= q3 + 1.5 * iqr)


    # q1 <- quantile(df[[col]], 0.25)
    # q3 <- quantile(df[[col]], 0.75)
    # iqr <- q3 - q1
    # # Mengaplikasikan aturan IQR untuk menyaring data
    # df <- filter(df, df[[col]] >= q1 - 1.5 * iqr, df[[col]] <= q3 + 1.5 * iqr)
    return(df)
}

for (col in colnames(df)){
    df <- remove_outlier(df, col)
}


# Membuat plot untuk setiap kolom
p1 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = diagnosis), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p2 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = radius_mean), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p3 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = texture_mean), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p4 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = perimeter_mean), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p5 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = area_mean), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p6 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = smoothness_mean), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p7 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = compactness_mean), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p8 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concavity_mean), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p9 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concave.points_mean), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p10 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = symmetry_mean), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p11 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fractal_dimension_mean), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p12 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = radius_se), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p13 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = texture_se), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p14 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = perimeter_se), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p15 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = area_se), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p16 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = smoothness_se), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p17 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = compactness_se), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p18 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concavity_se), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p19 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concave.points_se), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p20 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = symmetry_se), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p21 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fractal_dimension_se), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p22 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = radius_worst), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p23 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = texture_worst), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p24 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = perimeter_worst), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p25 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = area_worst), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p26 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = smoothness_worst), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p27 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = compactness_worst), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p28 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concavity_worst), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p29 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concave.points_worst), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p30 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = symmetry_worst), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p31 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fractal_dimension_worst), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()

# Menggabungkan plot menggunakan patchwork
plot <- p1 + p2 + p3 + p4 + p5 + p6 + p7 + p8 + p9 + p10 + 
        p11 + p12 + p13 + p14 + p15 + p16 + p17 + p18 + p19 + p20 + 
        p21 + p22 + p23 + p24 + p25 + p26 + p27 + p28 + p29 + p30 + p31
print(plot)

print(head(df))
numvars <- colnames(dplyr::select(df, -diagnosis))
scaler <- preProcess(df[numvars], method = "range")
df[numvars] <- predict(scaler, df[numvars])
print(head(df))

x <- subset(df, select = -diagnosis)
y <- as.factor(df$diagnosis)



set.seed(42)
# index <- createDataPartition(y, p = 0.7, list = TRUE)
# print(head(index))


# x_train <- x[index$Resample1, ]
# x_test <- x[-index$Resample1, ]
# y_train <- y[index$Resample1]
# y_test <- y[-index$Resample1]


split <- sample.split(y, SplitRatio = 0.7)
# x_train <- x[split, ]
# x_test <- x[!split, ]
# y_train <- y[split]
# y_test <- y[!split]

# Membuat data pelatihan
x_train <- subset(x, split == TRUE)
y_train <- subset(y, split == TRUE)
# Membuat data pengujian
x_test <- subset(x, split == FALSE)
y_test <- subset(y, split == FALSE)


ctrl <- trainControl(method="cv", number=5)
modelrf <- train(x_train, y_train, method="rf", trControl=ctrl)
print('accuracy mean')
print(modelrf$result)
print(mean(modelrf$result$Accuracy))
cat('\n')

model <- train(x_train, y_train, method="nnet", trControl=ctrl)
print('accuracy mean')
print(model$result)
print(mean(model$result$Accuracy))
cat('\n')

model <- train(x_train, y_train, method="rpart", trControl=ctrl)
print('accuracy mean')
print(model$result)
print(mean(model$result$Accuracy))
cat('\n')

model <- train(x_train, y_train, method="nb", trControl=ctrl)
print('accuracy mean')
print(model$result)
print(mean(model$result$Accuracy))
cat('\n')


y_pred_train <- predict(modelrf, x_train)
df_pred <- data.frame(
    y_train = y_train,
    y_pred_train = y_pred_train
)
print(head(df_pred))

df_pred <- filter(df_pred, y_train != y_pred_train)
print(head(df_pred))

# Calculate accuracy
accuracy <- confusionMatrix(data = y_pred_train, reference = y_train)$overall['Accuracy']
print(accuracy)

cat('\n\n\n\n')
# Calculate confusion matrix
print('Confusion Matrix')
cm <- confusionMatrix(data = y_pred_train, reference = y_train)
print(cm)




y_pred_test <- predict(modelrf, x_test)
df_pred <- data.frame(
    y_test = y_test,
    y_pred_test = y_pred_test
)
print(head(df_pred))

df_pred <- filter(df_pred, y_test != y_pred_test)
print(head(df_pred))

# Calculate accuracy
accuracy <- confusionMatrix(data = y_pred_test, reference = y_test)$overall['Accuracy']
print(accuracy)

cat('\n\n\n\n')
# Calculate confusion matrix
print('Confusion Matrix')
cm <- confusionMatrix(data = y_pred_test, reference = y_test)
print(cm)
